### PR TITLE
Change to new C++ based addon interface style

### PR DESCRIPTION
**Pull request by Rechi**
This are changes related to Kodi request xbmc/xbmc#12262

@AlwinEsch please check if the new interface is used correct.

### CODE REVIEW

File: src/maindx.cpp [hunk 1]
```diff
@@ -808,13 +828,10 @@ extern "C" void Render()
...
 
-extern "C" void Start(int iChannels, int iSamplesPerSec, int iBitsPerSample, const char* szSongName)
-{
-  samplesPerSec = iSamplesPerSec;
-}
-
-extern "C" void Stop()
+bool CVisualizationShadertoy::Start(int channels, int samplesPerSec, int bitsPerSample, std::string songName)
 {
+  samplesPerSec = samplesPerSec;
```
Comment (AlwinEsch) [inline src/maindx.cpp]: Maybe this match not the right destination?

File: src/maindx.cpp [hunk 2]
```diff
@@ -879,154 +896,90 @@ extern "C" void AudioData(const float* pAudioData, int iAudioDataLength, float *
...
-      LogActionString("VIS_ACTION_UPDATE_ALBUMART", (const char *)param);
-      break;
-    case VIS_ACTION_UPDATE_TRACK:
-      LogTrack((VisTrack *)param);
-      break;
-
-    default:
-      break;
-  }
+  LogAction("VIS_ACTION_LOAD_PRESET"); // TODO param is int *
```
Comment (AlwinEsch) [inline src/maindx.cpp]: What you mean with TODO?

File: src/maindx.cpp [hunk 3]
```diff
@@ -1063,49 +1016,19 @@ extern "C" void ADDON_Destroy()
...
-{
-  return ADDON_STATUS_OK;
-}
-
 //-- SetSetting ---------------------------------------------------------------
 // Set a specific Setting value (called from XBMC)
 // !!! Add-on master function !!!
 //-----------------------------------------------------------------------------
-extern "C" ADDON_STATUS ADDON_SetSetting(const char *strSetting, const void* value)
+ADDON_STATUS CVisualizationShadertoy::SetSetting(const std::string& settingName, const kodi::CSettingValue& settingValue)
```
Comment (AlwinEsch) [inline src/maindx.cpp]: By the use of `g_currentPreset = kodi::GetSettingInt("lastpresetidx");` and only the from addon self created value is this function no more really needed.

**Comment by AlwinEsch** [comment]
It looks OK, only is a `kodi::SetSettingInt("lastpresetidx", value);` somewhere on destruction missing.

**Comment by AlwinEsch** [comment]
Funny in original was also never the preset stored after usage :). Then it has never worked for what it was :D

File: src/maindx.cpp (hunk 2 above)
Comment (Rechi) [inline src/maindx.cpp]: it was there before, will remove it

File: src/maindx.cpp (hunk 3 above)
Comment (Rechi) [inline src/maindx.cpp]: whole `SetSetting` function can be removed?

Comment (AlwinEsch) [inline src/maindx.cpp]: Yes, is in settings.xml only declared as `<setting id="lastpresetidx" type="number" option="int" default="1" visible="false" />` with visible as false and can never changed by the user where SetSetting is needed.

**Comment by Rechi** [comment]
updated, I think this is ready to get merged now

**Pull request merged**